snuba/web/rpc/common.py

```diff
@@ -130,6 +141,83 @@ def attribute_key_to_expression(attr_key: AttributeKey) -> Expression:
     )
 
 
+def apply_virtual_columns(
+    query: Query, virtual_column_contexts: list[VirtualColumnContext]
+) -> None:
+    """Injects virtual column mappings into the clickhouse query. Works with NORMALIZED_COLUMNS on the table or
+    dynamic columns in attr_str
+
+    attr_num not supported because mapping on floats is a bad idea
+
+    Example:
+
+        SELECT
+          project_name AS `project_name`,
+          attr_str['release'] AS `release`,
+          attr_str['sentry.sdk.name'] AS `sentry.sdk.name`,
+        ... rest of query
+
+        contexts:
+            [   {from_column_name: project_id, to_column_name: project_name, value_map: {1: "sentry", 2: "snuba"}} ]
+
+
+        Query will be transformed into:
+
+        SELECT
+        -- see the project name column transformed and the value mapping injected
+          transform( CAST( project_id, 'String'), array( '1', '2'), array( 'sentry', 'snuba'), 'unknown') AS `project_name`,
+        --
+          attr_str['release'] AS `release`,
+          attr_str['sentry.sdk.name'] AS `sentry.sdk.name`,
+        ... rest of query
+
+    """
+
+    if not virtual_column_contexts:
+        return
+
+    mapped_column_to_context = {c.to_column_name: c for c in virtual_column_contexts}
+
+    def transform_expressions(expression: Expression) -> Expression:
+        # virtual columns will show up as `attr_str[virtual_column_name]` or `attr_num[virtual_column_name]`
+        if not isinstance(expression, SubscriptableReference):
+            return expression
+
+        if not expression.column != "attr_str":
+            return expression
+        context = mapped_column_to_context.get(str(expression.key.value))
+        attr_str = NestedColumn("attr_str")
+        if context:
+            if context.from_column_name in NORMALIZED_COLUMNS:
```

Comment:
instead of special casing this, can't you do like:

`f.transform(f.CAST(attribute_key_to_expression(request_name, String), 'String', ...)`

Reply:
yes good point